[Koh Jia Xian] iP

koh-jx opened the pull request:
> When the Duke is Sus!!

Duke Imposter is a management tool that tracks your tasks in an easily accessible and retrievable manner

- [x] Electrical: Download Data
- [x] Admin: Fix Wiring
- [ ] Reactor Meltdown /by 31/12/2021

## Features

1. Add or Remove tasks to your liking, or mark a current task as done
2. Specify the type of tasks to add
   * To-dos
   * Deadlines
   * Events
3. Find for a certain task by name or date (or both!)
4. **Fun text-based UI!**

*Best of all, it is FREE FOR USE! No subscription needed. **Donations are appreciated though** :) $50 or so wouldn't hurt*

## How to Use

1. Download the latest release [here](https://github.com/koh-jx/ip/releases)
2. Run the .jar file
3. Start Duking! ❤️❤️ 
   * Type help for the current list of commands
   * Type gubbai to terminate the program

### For Java developers

1. Download the source code.
2. Navigate to root/src/main/java/duke/Duke.java
3. Start the app via the main method

```java
// Duke.java
public static void main(String[] args) {
    Duke duke = new Duke();
    duke.start();
}
```

koh-jx committed:
Edited introductory text

koh-jx committed:
Listener

koh-jx committed:
Simple display

koh-jx committed:
Add, List

koh-jx committed:
abstract tasklist

koh-jx committed:
Add done regex

koh-jx committed:
Mark as Done, refactor

koh-jx committed:
Task refactor, Todo, Deadline

koh-jx committed:
Event

koh-jx committed:
change printSize(), enable regex case-insensitivity

koh-jx committed:
Refactor

koh-jx committed:
Textual changes, runtest.bat

koh-jx committed:
change display when adding task

koh-jx committed:
IndexOutOfBounds exception

koh-jx committed:
Refactor

koh-jx committed:
Delete tasks

koh-jx committed:
minor formatting changes

koh-jx committed:
Refactor Display to Ui

koh-jx committed:
add javadocs

koh-jx committed:
Save task

koh-jx committed:
load tasks, overload task(list) constructors

koh-jx committed:
Refactor to Storage, error handling

koh-jx committed:
Relative path

koh-jx committed:
Localdate integration

koh-jx committed:
Merge branch 'branch-Level-7'

koh-jx committed:
Cleanup after merge

koh-jx committed:
Fix merge conflicts

koh-jx committed:
Add find date, cleanup

koh-jx committed:
Add find to list as optional argument

koh-jx committed:
Package duke and task

koh-jx committed:
Cleanup

koh-jx committed:
Delete tasklist.txt

koh-jx committed:
gitignore of tasklist.txt

koh-jx committed:
Merge branch 'master' of https://github.com/koh-jx/ip

koh-jx committed:
Add Junit5 tests

koh-jx committed:
Create directory for save file if does not exist

koh-jx committed:
Add Javadocs

koh-jx committed:
Tidy up code

koh-jx committed:
List arguments, chain args

koh-jx committed:
Fix merge conflicts

koh-jx committed:
Fix merge conflicts

koh-jx committed:
Fix merge conflicts

koh-jx committed:
Fix merge conflicts

koh-jx committed:
Cleanup

koh-jx committed:
Change access modifiers

koh-jx committed:
Refactor Commands

koh-jx committed:
Add Command Javadocs

koh-jx committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/DukeParser.java

koh-jx committed:
Tidy help function

koh-jx committed:
Add Gradle

koh-jx committed:
Revert "Add Gradle"

This reverts commit 88e80c5c0992a6dda8a4e00933ce7b7940570fbe.

koh-jx committed:
Update README.md

koh-jx committed:
Add Gradle, remove unused imports

koh-jx committed:
Merge branch 'master' of https://github.com/koh-jx/ip

koh-jx committed:
checkstyleMain

koh-jx committed:
checkstyleTest

koh-jx committed:
Update README.md

koh-jx committed:
Setup JavaFX

koh-jx committed:
Parts 1-3

jeffsieu commented on README.md:
```diff
@@ -1,24 +1,69 @@
+# There is 1️⃣ Imposter among us ඞ
 
-This is a project template for a greenfield Java project. It's named after the Java mascot _Duke_. Given below are instructions on how to use it.
+> When the Duke is Sus!!
```
what a boomer

koh-jx replied:
what a zoomer

jeffsieu commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,40 @@
+
+    /**
+     * Constructor
+     * Instantiates a (saved) task.Task List and a duke.DukeListener for duke.Duke
```
Could use `{@link <ClassName>}` to allow JavaDoc to link the class in the generated docs.

koh-jx replied:
Wow, I didn't know this! Thanks for the suggestion! 😜

jeffsieu commented on src/main/java/duke/DukeListener.java:
```diff
@@ -0,0 +1,50 @@
+
+
+/**
+ * Waits for and processes a User Input
```
"User Input" could be "user input". No reason to capitalize here.

koh-jx replied:
You are right!

jeffsieu commented on src/main/java/duke/DukeListener.java:
```diff
@@ -0,0 +1,50 @@
+ * Waits for and processes a User Input
+ */
+public class DukeListener {
+    private final Scanner sc;
```
No harm calling it `scanner` too. Would be clearer in parts of code that uses this variable.

koh-jx replied:
You mentioned a very good point!

jeffsieu commented on src/main/java/duke/DukeParser.java:
```diff
@@ -0,0 +1,85 @@
+            Pattern.CASE_INSENSITIVE);
+
+    /**
+     * Constructor; instantiates the task list to be edited
```
First word after semi-colon should be capitalized.

koh-jx replied:
Ah totally missed this out! Thanks

jeffsieu commented on src/main/java/duke/DukeParser.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * Patterns for the Parser to look out for in the input
+     */
+    private final Pattern listPattern = Pattern.compile("list( .+)?", Pattern.CASE_INSENSITIVE);
```
![image](https://user-images.githubusercontent.com/8487294/131513743-24d8d949-6d5f-4bfd-8323-1fd7c8631317.png)
Could be in UPPERCASE, and written like `PATTERN_LIST`, `PATTERN_DONE`, etc.

jeffsieu replied:
No reason not to be static too.

koh-jx replied:
Ah I missed this out as well, thanks

jeffsieu commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,33 @@
+
+    public abstract void execute();
+
+    protected static LocalDate getDate(String date) throws DateTimeParseException {
```
Not necessarily a Command-specific method. Could potentially be moved to a Utils class.

koh-jx replied:
I didn't have a use for a Utils class prior to making this function, so this was the most relevant place to put it. However, you raise a very good point here

jeffsieu commented on src/main/java/duke/command/CommandAddDeadline.java:
```diff
@@ -0,0 +1,46 @@
+     * @param taskList Task list
+     * @param groups Matcher with info needed to create the class
+     */
+    public CommandAddDeadline(TaskList taskList, Matcher groups) {
```
`groups` could be abstracted out, or arguments can be extracted out into an array before passing into the Command constructor.

koh-jx replied:
That's true!

jeffsieu commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+     * @param task Task to add
+     */
+    public void add(Task task) {
+        Optional.ofNullable(task)
```
Really obscure way of doing a null check haha

koh-jx replied:
:P

jeffsieu commented on src/test/java/duke/DukeParserTest.java:
```diff
@@ -0,0 +1,26 @@
+    @Test
+    void parseInput_randomString_nothingParsed() {
+
+
```
Wow two empty lines.

jeffsieu reviewed (commented):
Well done! Great effort!

koh-jx committed:
Implement JavaFX with .fxml

koh-jx committed:
Implement Duke in JavaFX

nicole-luo-exe commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,32 @@
+     * @param input String input from user
+     */
+    protected String getResponse(String input) {
+        // TaskList related inputs
```
Is there a reason this comment is not within but separate from the header comment?

koh-jx replied:
Oh, this was from previous commits where I initially handled non-tasklist related inputs within Duke.java itself. It's an oversight, thankss

nicole-luo-exe commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,51 @@
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+/**
```
A newline might be missing between lines 9-10.

nicole-luo-exe commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,107 @@
+     * Converts a given string for the .txt save file to a valid Task
+     *
+     * @param task String from .txt save file
+     * @return task.Task corresponding to the string
```
You might have mistakenly entered a period instead of a space here. Otherwise, comments are clear and easy to understand. 😃

nicole-luo-exe commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,107 @@
+        String[] args = task.split("\\t");
+        String taskType = args[0];
+        try {
+            switch(taskType) {
```
Should there be a space after switch?

koh-jx replied:
Ah, yes, thanks

nicole-luo-exe commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Exit text
+     */
+    public static String goodBye() {
```
Perhaps change the method name to a verb like getGoodbyeMsg?

nicole-luo-exe replied:
Similar issue in line 14.

nicole-luo-exe commented on src/main/java/duke/command/CommandAddEvent.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    /**
+     * Add an event to the task list if valid
```
Header comments for methods should be in the form Adds instead of Add.

nicole-luo-exe replied:
Maybe check all of your CommandAdd*.java files with this header comment? Apart from this, everything else looks fine coding-standard-wise. 😸

nicole-luo-exe commented on src/main/java/duke/command/CommandAddTodo.java:
```diff
@@ -0,0 +1,38 @@
+     * Constructor
+     *
+     * @param taskList Task list to use
+     * @param desc description of task
```
Seems like most of your elaborations following the param name is capitalized so you might want to capitalize "description" to keep it standardized?

nicole-luo-exe commented on src/main/java/duke/command/CommandHelp.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    /**
+     * Display all commands
```
Perhaps change display to displays? [source](https://se-education.org/guides/conventions/java/intermediate.html#comments)

nicole-luo-exe replied:
I noticed the same issue in several other places too (CommandList.java line 63, CommandInvalid.java line 25- for your easy reference)

koh-jx replied:
Thank you! Sorry for the trouble :')

nicole-luo-exe commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,68 @@
+        this.isDone = isDone;
+    }
+
+    /**Toggles completion of Task
```
"Toggles..." should be in newline from /**. Likely a small oversight.

nicole-luo-exe commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * String representation of Task
+     *
+     * @return task display
```
You might want to keep the phrase after @return standardized in terms of capitalization (i.e. either all to start with a capital or lowercase letter).

nicole-luo-exe commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,126 @@
+     */
+    public String toggleDone(int index) {
+        try {
+            boolean result = tasks.get(index - 1).toggleDone();
```
Booleans should be named to sound like booleans.

nicole-luo-exe commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,126 @@
+            return "There is nothing to display! :angery:";
+        } else {
+            // Get resultant list after applying all filters
+            List<Task> list = tasks.stream()
```
Perhaps filteredTasks for the name instead? Names should provide information about the objects they hold and plural form should be used on names representing a collection of objects. [source](https://se-education.org/guides/conventions/java/intermediate.html#naming)

nicole-luo-exe reviewed (commented):
Generally code does abide by Java coding standards but there are small oversights here and there. Overall, well done! Comments are clear and names are for the most part appropriate.
👍

KengXIII commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,126 @@
+     */
+    public String delete(int index) {
+        try {
+            Task removed = tasks.remove(index - 1);
```
Would 'removedTask' be a less vague variable name here?

KengXIII commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * Get response from duke from the GUI
+     *
+     * @param input String input from user
```
parameter description should follow with a punctuation

KengXIII commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    /**
+     * Exit text
```
The JavaDocs for this method could be a little more descriptive and follow the form "Prints..."/ "Shows..."

koh-jx committed:
Fix errors from PR comments

koh-jx committed:
Merge branch 'branch-level-10'

koh-jx committed:
Cleanup JavaFX, Fix bugs

Placed relevant JavaFX windows and java files into their respective
packages; ui and util packages.

Fix bugs where text was being wrapped in the UI, and intro message
not being displayed upon startup.

koh-jx committed:
Merge branch 'refactor-javafx'

koh-jx committed:
Improve coding quality, fix checkstyle violations

Changed certain functions and variables to meet coding quality
standards. Most notable, the shift of all output strings to Ui.java
removes the presence of magic strings in the program.

koh-jx committed:
Add assertions

Assertions added and enabled to ensure the validity of variable states

koh-jx committed:
Merge pull request #1 from koh-jx/branch-A-CodeQuality

Improve coding quality, fix checkstyle violations

koh-jx committed:
Fix merge conflicts

koh-jx committed:
Merge pull request #2 from koh-jx/branch-A-Assertions

Add assertions

koh-jx committed:
Update gradle version, add shadowJar

koh-jx committed:
Change GUI

koh-jx committed:
Cleanup javadocs

koh-jx committed:
Set theme jekyll-theme-merlot

koh-jx committed:
Add minor UI changes

koh-jx committed:
Merge branch 'master' of https://github.com/koh-jx/ip

koh-jx committed:
Add user guide

koh-jx committed:
Edit user guide

koh-jx committed:
Touch-up user guide

koh-jx committed:
Set theme jekyll-theme-midnight

koh-jx committed:
Add manifest

koh-jx committed:
Merge branch 'master' of https://github.com/koh-jx/ip

koh-jx committed:
Add Sort functionality

koh-jx committed:
Add sorting features

Added sorting by completion status and date added, which uses
LocalDateTime to compare dates. Also implemented the date added to
Storage functions, and tweaked formatting of commands. Edited
help command and user guide.

koh-jx committed:
Fix Javadocs, userguide

koh-jx committed:
Touchup javadocs

koh-jx committed:
Update README.md

koh-jx committed:
Update README.md

koh-jx committed:
Update README.md
